test/utils/wasm3_engine.cpp

```diff
@@ -111,24 +144,54 @@ fizzy::bytes_view Wasm3Engine::get_memory() const
 }
 
 std::optional<WasmEngine::FuncRef> Wasm3Engine::find_function(
-    std::string_view name, std::string_view) const
+    std::string_view name, std::string_view signature) const
 {
     IM3Function function;
-    if (m3_FindFunction(&function, m_runtime, name.data()) == m3Err_none)
-        return reinterpret_cast<WasmEngine::FuncRef>(function);
-    return std::nullopt;
+    if (m3_FindFunction(&function, m_runtime, name.data()) != m3Err_none)
+        return std::nullopt;
+
+    std::vector<M3ValueType> inputs;
+    std::vector<M3ValueType> outputs;
+    std::tie(inputs, outputs) = translate_signature(signature);
+
+    if (inputs.size() != m3_GetArgCount(function))
+        return std::nullopt;
+    for (unsigned i = 0; i < m3_GetArgCount(function); i++)
+        if (inputs[i] != m3_GetArgType(function, i))
+            return std::nullopt;
+
+    if (outputs.size() != m3_GetRetCount(function))
+        return std::nullopt;
+    for (unsigned i = 0; i < m3_GetRetCount(function); i++)
+        if (outputs[i] != m3_GetRetType(function, i))
+            return std::nullopt;
+
+    return reinterpret_cast<WasmEngine::FuncRef>(function);
 }
 
 WasmEngine::Result Wasm3Engine::execute(
     WasmEngine::FuncRef func_ref, const std::vector<uint64_t>& args)
 {
-    unsigned ret_valid;
-    uint64_t ret_value;
     IM3Function function = reinterpret_cast<IM3Function>(func_ref);
-    auto const result = m3_CallProper(
-        function, static_cast<uint32_t>(args.size()), args.data(), &ret_valid, &ret_value);
-    if (result == m3Err_none)
-        return {false, ret_valid ? ret_value : std::optional<uint64_t>{}};
+
+    std::vector<const void*> argPtrs;
```

Comment:
If you want ultimate speed you can use stack-allocated fixed-size array. Just bail out if too much arguments provided.

Reply:
```
fizzy/test/utils/wasm3_engine.cpp:147:24: error: variable length arrays are a C99 feature [-Werror,-Wvla-extension]
    const void* argPtrs[args.size()];
```
 
Or not sure what you've meant?

Reply:
Ah you meant like allocating an array with like 32 items, and fail if `args.size()` is larger.

Reply:
Not sure if this makes a lot of speed difference, but pushed a version.